Enh/classif det

stephlazar opened the pull request:
Improves classify_shots of the UHFQC_classifier_detector:

- raises errors when combination of boolean flags does not make sense

- adds option to do ro_corrected_stored_mtx or ro_corrected_seq_cal_mtx. **This means that now you do not have to set the qubit parameter acq_state_prob_mtx to the identity in order to disable RO correction in the detector. The readout correction is not entirely controlled by the two new flags.**

ro_corrected_stored_mtx : corrects with the corresponding qubit matrix stored in acq_state_prob_mtx 

ro_corrected_seq_cal_mtx: corrects with state probability matrix calculated from the averaged calibration states at the end of each sequence.

The changes to the detector function proposed in this pull request were tested both on a virtual setup and on the XLD physical setup.

The changes proposed in this pull request should not break existing code, and users should not need to change their workflow other than the statement above in **bold**.

@chellings @nathlacroix be aware of this pull request. Your feedback is of course welcome.

Commit:
Improved classify_shots of the UHFQC_classifier_detector:
 - raises errors when combination of boolean flags does not make sense
 - adds option to do ro_corrected_stored_mtx or ro_corrected_seq_cal_mtx.

Commit:
Adds new UHFQC_classifier_detector boolean flags for processing data to the det_get_values_kws dictionary
 in measure_cphase and measure_two_qubit_randomized_benchmarking. Also adds det_get_values_kws as input
 parameter to these measure functions.

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

antsr commented on pycqed/measurement/detector_functions.py:
```diff
@@ -1359,54 +1359,84 @@ def process_data(self, data_raw):
 
     def classify_shots(self, data, classifier_params_list,
-                       state_prob_mtx_list=None, averaged=False,
-                       thresholded=True):
+                       averaged=True, thresholded=False,
+                       ro_corrected_stored_mtx=False,
+                       ro_corrected_seq_cal_mtx=False,
+                       state_prob_mtx_list=None, **kw):
```
Could you add a docstring explaining the meaning of all the parameters of this function.

stephlazar replied:
I now added explicit docstring.

antsr commented on pycqed/measurement/detector_functions.py:
```diff
@@ -1317,14 +1317,14 @@ def prepare(self, sweep_points):
+        self.UHFQC.qudev_acquisition_initialize(
+            channels=self.channels, samples=self.nr_shots*self.nr_sweep_points,
+            averages=1, #for single shot readout
+            loop_cnt=int(self.nr_shots), mode='rl')
```
I guess it is not the scope of this PR, but the init here is very similar to the integration_logging_detector. This class should maybe inherit from the logging_det to clean up the code.

stephlazar replied:
I made this class inherit from the UHFQC_integration_logging_det and cleaned it up a bit.

antsr reviewed (changes requested):
Could you please update the docstring of the class itself and add docstrings to some of the more complex intermediate functions.

This class is really hard to read for me as it is now. Where different parameters are passed in / defined could be improved for this class in general. Example: The split between process_data and classify_shots seems a bit arbitrary. Some of the initial processing is done in process_data, but then rest of it is done in classify shots.

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Significantly cleans up the UHFQC_classifier_detector and adds explicit docstring.

stephlazar commented:
@antsr I added explicit docstring, I cleaned up the class and made it inherit from the UHFQC_integration_logging_det.
However, now I think we should first test that everything still works in a measurement on the setup.

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

antsr commented:
I'll try and test this on XLD this week to get it merged.

Commit:
Merge branch 'qudev_master' into enh/classif_det

stephlazar commented:
Sounds good! 👍

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

# Conflicts:
#	pycqed/measurement/multi_qubit_module.py

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

# Conflicts:
#	pycqed/measurement/detector_functions.py

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Bugfixes and improvements to MC, UHFQC_integration_logging_det, UHFQC_classifier_detector, UHFQC_multi_detector.

- adds the fixes required to make SSRO measurements work correctly!

stephlazar commented:
I added even more improvements

- [x] Was tested on virtual setup

- [x] Should still be tested on a real setup

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

# Conflicts:
#	pycqed/measurement/detector_functions.py
#	pycqed/measurement/measurement_control.py

Commit:
Merge branch 'qudev_master' into enh/classif_det

# Conflicts:
#	pycqed/measurement/multi_qubit_module.py

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
analysis_toolbox.py/predict_gm_proba_from_clf: Reshape array if it is 1D into a 2D.

Commit:
UHFQC_classifier_detector: upgrades to work for qubits as well

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
UHFQC_classifier_detector: fix bug that made it crash when classified == False

Commit:
Merge branch 'qudev_master' into enh/classif_det

Commit:
UHFQC_classifier_detector: fixes bug in default value for classifier_params and for n_meas_objs in init

Commit:
QuDev_transmon: fixes bug in update_detector_functions where classifier params and state prob mtx were not passed as lists to UHF_classifier_detector

antsr reviewed (approved):
Looks good to me now.
